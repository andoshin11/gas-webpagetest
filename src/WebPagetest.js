import Utils from './Utils'

export default class WebPagetest {
  server
  key

  /**
   * @param {type}	key - runTest するときは必須。
   * @param {type}	server - 省略すれば www.webpagetest.org
   */
  constructor(key, server) {
    this.key = key
    this.server = server || 'https://www.webpagetest.org'
  }

  /**
   * WebPagetest 実行
   *
   * @param {type}	url - this is the parameter url
   * @param {type}	options - this is the parameter options
   *
   * @return {} testId
   */
  runTest(url, options) {
    const requestURL = this.generateRunTestURL(url, options)
    const {data: {testId}} = Utils.fetchRequest(requestURL)

    return testId
  }

  /**
   * WebPagetest 実行用 URL 生成
   *
   * @param {type}	url - 必須。
   * @param {type}	options - 省略可能。
   *
   * @return {String} URL
   */
  generateRunTestURL(url, options = {}) {
    return `${this.server}/runtest.php`
      + `?url=${encodeURIComponent(url)}`
      + `&location=${options.location || 'ec2-ap-northeast-1.3GFast'}`
      + `&fvonly=${options.fvonly || 1}`
      + `&video=${options.video || 1}`
      + `&f=${options.format || 'JSON'}`
      + `&k=${this.key}`
      + `&mobile=${options.mobile || 1}`
      + `&mobileDevice=${options.mobileDevice || 'Nexus7'}`
      + `&lighthouse=${options.lighthouse || 1}`
  }

  /**
   * テスト結果取得
   *
   * @param {type}	testId - this is the parameter testId
   *
   * @return {Object} responsedata
   */
  getTestResults(testId) {
    const requestURL = this.generateTestResultsURL(testId)
    const {data} = Utils.fetchRequest(requestURL)

    return [
      data.completed,
      Utilities.formatDate(new Date(data.completed * 1000), 'GMT+9', 'yyyyMMddHH'),
      Utils.convertToSeconds(data.average.firstView.TTFB),
      Utils.convertToSeconds(data.average.firstView.loadTime),
      Utils.convertToSeconds(data.average.firstView.domInteractive),
      Utils.convertToSeconds(data.average.firstView.render),
      Utils.convertToSeconds(data.average.firstView.visualComplete),
      Utils.convertToSeconds(data.average.firstView.fullyLoaded),
      Utils.convertToSeconds(data.average.firstView.firstPaint),
      Utils.convertToSeconds(data.average.firstView.SpeedIndex),
      data.summary,
    ]
  }

  /**
   * WebPagetest 実行結果用 URL 生成
   *
   * @param {type}	testId - this is the parameter testId
   *
   * @return {String} URL
   */
  generateTestResultsURL(testId) {
    return `${this.server}/jsonResult.php?test=${testId}&pagespeed=1`
  }
}
