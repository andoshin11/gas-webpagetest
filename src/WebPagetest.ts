import Utils = require('./Utils')
import 'core-js/modules/es6.array.fill'

class WebPagetest {
  /**
   * @param {type}	key - runTest するときは必須。
   * @param {type}	server - 省略すれば www.webpagetest.org
   */
  constructor(
    private key?: string,
    private server: string = 'https://www.webpagetest.org',
  ) {}

  /**
   * WebPagetest 実行
   *
   * @param {type}	url - this is the parameter url
   * @param {type}	options - this is the parameter options
   *
   * @return {} testId
   */
  public runTest(url: string, options?: Options): string {
    const requestURL = this.generateRunTestURL(url, options)
    const {
      data: { testId },
    } = Utils.fetch(requestURL)

    return testId
  }

  public getTestStatus(testId: string): number {
    const {
      data: { statusCode },
    } = Utils.fetch(this.generateTestStatusURL(testId))

    return statusCode
  }

  /**
   * テスト結果取得
   *
   * @param {type}	testId - this is the parameter testId
   *
   * @return {Object} responsedata
   */
  public getTestResults(testId) {
    const statusCode = this.getTestStatus(testId)

    if (statusCode !== 200) {
      return new Array(27).fill('')
    }

    const requestURL = this.generateTestResultsURL(testId)
    const {
      data: {
        completed,
        median: {
          firstView: {
            TTFB: firstByte,
            firstLayout,
            firstPaint,
            firstContentfulPaint,
            firstMeaningfulPaint,
            SpeedIndex: speedIndex,
            domInteractive,
            loadTime,
            visualComplete,
            fullyLoaded,
            'lighthouse.Performance.interactive': timeToInteractive,
            breakdown,
          },
        },
        summary,
      },
    } = Utils.fetch(requestURL)
    const date = new Date(completed * 1000)

    return [
      completed,
      Utilities.formatDate(date, 'GMT+9', 'yyyyMMddHH'),
      Utilities.formatDate(date, 'GMT+9', 'yyyyMMdd'),
      Utils.transform(firstByte),
      Utils.transform(firstLayout),
      Utils.transform(firstPaint),
      Utils.transform(firstContentfulPaint),
      Utils.transform(firstMeaningfulPaint),
      Utils.transform(speedIndex),
      Utils.transform(domInteractive),
      Utils.transform(loadTime),
      Utils.transform(visualComplete),
      Utils.transform(fullyLoaded),
      Utils.transform(timeToInteractive),
      Utils.transform(breakdown.html.bytes, 1),
      breakdown.html.requests,
      Utils.transform(breakdown.js.bytes, 1),
      breakdown.js.requests,
      Utils.transform(breakdown.css.bytes, 1),
      breakdown.css.requests,
      Utils.transform(breakdown.image.bytes, 1),
      breakdown.image.requests,
      Utils.transform(breakdown.font.bytes, 1),
      breakdown.font.requests,
      Utils.transform(breakdown.other.bytes, 1),
      breakdown.other.requests,
      summary,
    ]
  }

  public test = this.runTest
  public results = this.getTestResults

  /**
   * WebPagetest 実行用 URL 生成
   *
   * @param {type}	url - 必須。
   * @param {type}	options - 省略可能。
   *
   * @return {String} URL
   */
  private generateRunTestURL(url: string, options: Options = {}): string {
    return (
      `${this.server}/runtest.php` +
      `?url=${encodeURIComponent(url)}` +
      `&location=${options.location || 'ec2-ap-northeast-1.3GFast'}` +
      `&runs=${options.runs || 1}` +
      `&fvonly=${options.fvonly || 1}` +
      `&video=${options.video || 1}` +
      `&f=${options.format || 'JSON'}` +
      `&noopt=${options.noOptimization || 1}` +
      `&k=${this.key}` +
      `&mobile=${options.mobile || 1}` +
      `&mobileDevice=${options.mobileDevice || 'Pixel'}` +
      `&lighthouse=${options.lighthouse || 1}`
    )
  }

  private generateTestStatusURL(testId: string) {
    return `${this.server}/testStatus.php?f=json&test=${testId}`
  }

  /**
   * WebPagetest 実行結果用 URL 生成
   *
   * @param {type}	testId - this is the parameter testId
   *
   * @return {String} URL
   */
  private generateTestResultsURL(testId: string) {
    return `${this.server}/jsonResult.php?test=${testId}&pagespeed=1`
  }
}

export = WebPagetest
