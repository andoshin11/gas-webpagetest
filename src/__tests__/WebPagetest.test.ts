import WebPagetest = require('../WebPagetest')

describe('WebPagetest', () => {
  describe('#generateRunTestURL', () => {
    it('should return test url when it has not options', () => {
      const webPagetest = new WebPagetest('key')
      const url = webPagetest.generateRunTestURL('https://example.com')
      expect(url).toMatchInlineSnapshot(
        `"https://www.webpagetest.org/runtest.php?url=https%3A%2F%2Fexample.com&location=ec2-ap-northeast-1.3GFast&runs=1&fvonly=1&video=1&f=JSON&noopt=1&k=key&mobile=1&mobileDevice=Pixel&lighthouse=1"`
      )
    })
    it('should return test url when it has options', () => {
      const webPagetest = new WebPagetest('key')
      const url = webPagetest.generateRunTestURL('https://example.com', {
        location: 'ec2-ap-northeast-1:Chrome',
        runs: 1,
        fvonly: 0,
        video: 0,
        mobile: 0,
        format: 'JSON',
        noOptimization: 0,
        mobileDevice: 'iPhone5c',
        lighthouse: 0,
      })
      expect(url).toMatchInlineSnapshot(
        `"https://www.webpagetest.org/runtest.php?url=https%3A%2F%2Fexample.com&location=ec2-ap-northeast-1%3AChrome&runs=1&fvonly=0&video=0&f=JSON&noopt=0&k=key&mobile=0&mobileDevice=iPhone5c&lighthouse=0"`
      )
    })
  })
})
