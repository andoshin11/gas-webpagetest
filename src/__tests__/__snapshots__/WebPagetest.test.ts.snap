// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebPagetest generateTestResult should return results: WebPagetest-response-google.com 1`] = `
Array [
  Object {
    "name": "completedTimeStamp",
    "value": 1532835031,
  },
  Object {
    "name": "yyyyMMddHH",
    "value": "2018-07-29T03:30:31.000Z, GMT+9, yyyyMMddHH",
  },
  Object {
    "name": "yyyyMMdd",
    "value": "2018-07-29T03:30:31.000Z, GMT+9, yyyyMMdd",
  },
  Object {
    "name": "firstByte",
    "value": "0.48",
  },
  Object {
    "name": "firstLayout",
    "value": "0.54",
  },
  Object {
    "name": "firstPaint",
    "value": "0.59",
  },
  Object {
    "name": "firstContentfulPaint",
    "value": "0.59",
  },
  Object {
    "name": "firstMeaningfulPaint",
    "value": "0.59",
  },
  Object {
    "name": "speedIndex",
    "value": "0.61",
  },
  Object {
    "name": "domInteractive",
    "value": "0.63",
  },
  Object {
    "name": "loadTime",
    "value": "1.47",
  },
  Object {
    "name": "visualComplete",
    "value": "1.10",
  },
  Object {
    "name": "fullyLoaded",
    "value": "1.91",
  },
  Object {
    "name": "timeToInteractive",
    "value": "1.96",
  },
  Object {
    "name": "html.bytes",
    "value": "68.6",
  },
  Object {
    "name": "html.requests",
    "value": 9,
  },
  Object {
    "name": "js.bytes",
    "value": "286.7",
  },
  Object {
    "name": "js.requests",
    "value": 5,
  },
  Object {
    "name": "css.bytes",
    "value": "0.0",
  },
  Object {
    "name": "css.requests",
    "value": 0,
  },
  Object {
    "name": "image.bytes",
    "value": "37.7",
  },
  Object {
    "name": "image.requests",
    "value": 6,
  },
  Object {
    "name": "font.bytes",
    "value": "0.0",
  },
  Object {
    "name": "font.requests",
    "value": 0,
  },
  Object {
    "name": "flash.bytes",
    "value": "0.0",
  },
  Object {
    "name": "flash.requests",
    "value": 0,
  },
  Object {
    "name": "video.bytes",
    "value": "0.0",
  },
  Object {
    "name": "video.requests",
    "value": 0,
  },
  Object {
    "name": "other.bytes",
    "value": "0.0",
  },
  Object {
    "name": "other.requests",
    "value": 0,
  },
  Object {
    "name": "summary",
    "value": "http://www.webpagetest.org/results.php?test=180729_D6_c92a1ce569da3ae435ac558fc934a38b",
  },
  Object {
    "name": "finalURL",
    "value": "https://www.google.com/",
  },
  Object {
    "name": "browserName",
    "value": "Chrome",
  },
  Object {
    "name": "browserVersion",
    "value": "68.0.3440.75",
  },
  Object {
    "name": "visualComplete85",
    "value": "0.6",
  },
  Object {
    "name": "visualComplete90",
    "value": "0.6",
  },
  Object {
    "name": "visualComplete95",
    "value": "0.7",
  },
  Object {
    "name": "visualComplete99",
    "value": "0.7",
  },
  Object {
    "name": "bytesOut",
    "value": "12.3",
  },
  Object {
    "name": "bytesOutDoc",
    "value": "8.7",
  },
  Object {
    "name": "bytesIn",
    "value": "393.0",
  },
  Object {
    "name": "bytesInDoc",
    "value": "374.2",
  },
  Object {
    "name": "connections",
    "value": 8,
  },
  Object {
    "name": "requestsDoc",
    "value": 15,
  },
  Object {
    "name": "responses_200",
    "value": 13,
  },
  Object {
    "name": "responses_404",
    "value": 0,
  },
  Object {
    "name": "responses_other",
    "value": 7,
  },
]
`;

exports[`WebPagetest generateTestResult should return results: WebPagetest-response-youtube.com 1`] = `
Array [
  Object {
    "name": "completedTimeStamp",
    "value": 1532877614,
  },
  Object {
    "name": "yyyyMMddHH",
    "value": "2018-07-29T15:20:14.000Z, GMT+9, yyyyMMddHH",
  },
  Object {
    "name": "yyyyMMdd",
    "value": "2018-07-29T15:20:14.000Z, GMT+9, yyyyMMdd",
  },
  Object {
    "name": "firstByte",
    "value": "0.27",
  },
  Object {
    "name": "firstLayout",
    "value": "2.93",
  },
  Object {
    "name": "firstPaint",
    "value": "3.11",
  },
  Object {
    "name": "firstContentfulPaint",
    "value": "3.11",
  },
  Object {
    "name": "firstMeaningfulPaint",
    "value": "3.99",
  },
  Object {
    "name": "speedIndex",
    "value": "3.93",
  },
  Object {
    "name": "domInteractive",
    "value": "3.61",
  },
  Object {
    "name": "loadTime",
    "value": "4.01",
  },
  Object {
    "name": "visualComplete",
    "value": "5.20",
  },
  Object {
    "name": "fullyLoaded",
    "value": "6.34",
  },
  Object {
    "name": "timeToInteractive",
    "value": "17.38",
  },
  Object {
    "name": "html.bytes",
    "value": "448.6",
  },
  Object {
    "name": "html.requests",
    "value": 11,
  },
  Object {
    "name": "js.bytes",
    "value": "1067.7",
  },
  Object {
    "name": "js.requests",
    "value": 13,
  },
  Object {
    "name": "css.bytes",
    "value": "52.2",
  },
  Object {
    "name": "css.requests",
    "value": 3,
  },
  Object {
    "name": "image.bytes",
    "value": "116.7",
  },
  Object {
    "name": "image.requests",
    "value": 21,
  },
  Object {
    "name": "font.bytes",
    "value": "31.3",
  },
  Object {
    "name": "font.requests",
    "value": 2,
  },
  Object {
    "name": "flash.bytes",
    "value": "0.0",
  },
  Object {
    "name": "flash.requests",
    "value": 0,
  },
  Object {
    "name": "video.bytes",
    "value": "432.5",
  },
  Object {
    "name": "video.requests",
    "value": 3,
  },
  Object {
    "name": "other.bytes",
    "value": "253.7",
  },
  Object {
    "name": "other.requests",
    "value": 6,
  },
  Object {
    "name": "summary",
    "value": "http://www.webpagetest.org/results.php?test=180729_3T_4e4850ed9a41879faafa4c55ee411fce",
  },
  Object {
    "name": "finalURL",
    "value": "https://www.youtube.com/watch?v=6UeRMMI_IzI",
  },
  Object {
    "name": "browserName",
    "value": "Chrome",
  },
  Object {
    "name": "browserVersion",
    "value": "68.0.3440.75",
  },
  Object {
    "name": "visualComplete85",
    "value": "5.0",
  },
  Object {
    "name": "visualComplete90",
    "value": "5.0",
  },
  Object {
    "name": "visualComplete95",
    "value": "5.2",
  },
  Object {
    "name": "visualComplete99",
    "value": "5.2",
  },
  Object {
    "name": "bytesOut",
    "value": "32.3",
  },
  Object {
    "name": "bytesOutDoc",
    "value": "14.0",
  },
  Object {
    "name": "bytesIn",
    "value": "2402.7",
  },
  Object {
    "name": "bytesInDoc",
    "value": "1632.8",
  },
  Object {
    "name": "connections",
    "value": 15,
  },
  Object {
    "name": "requestsDoc",
    "value": 27,
  },
  Object {
    "name": "responses_200",
    "value": 52,
  },
  Object {
    "name": "responses_404",
    "value": 0,
  },
  Object {
    "name": "responses_other",
    "value": 7,
  },
]
`;
